### ISSUE TITLE
getting loss value periodically [NoobQuestion#1]

### ISSUE THREAD

**Issue by IgorWojas** [mlpack/mlpack#3604]
Hello,

i'm not sure if this question belongs here or in the Ensmallen repo...

How to get loss value when training a model?

PrintLoss() and ProgressBar() do it consistently but i would like to get loss value every n-th epoch because printing anything in the console slows down the training.

so using the following should solve the problem
 
` if (epochs % 1000 == 0) { cout << "loss = " << lossVal << "\n"; }`

how do i get lossVal?

what i tried:
- model.Evaluate( ... )
- model.EvaluateWithGradient( ... )
These give something but its not loss.

I've also tried searching in the source code but this is far beyond my level of uderstanding.

Sorry for the noob question and thanks in advance!

EDIT: As this may be related to optimizer I'm using Adam for now and the network is FFN.

**Comment by Adarsh97** [comment]
my doubt is what model.Evaluate() does ?
I am using vae (with FFN) when i was using model.evaluate() i am getting a value (double), i thought it is error.
But it is giving negative values as well.

**Comment by IgorWojas** [comment]
@Adarsh97 - Doc says "used to monitor prograss while training" . This description led me to try it. When i used ens::PrintLoss() along with model.Evaluate() i could see that the values are different. Another thing i tried yesterday was to print the value that the Train() function returns (also double) . The doc says "the final objective of the trained model" which seems vague to me.
Please post a comment if you find anything. Cheers!!

**Comment by vibhorjoshi** [comment]
To obtain the loss value during training, you can use the Forward method to compute predictions and then calculate the loss using a criterion

**Comment by Adarsh97** [comment]
In that case what is the difference between predict and forward?

**Comment by rcurtin** [comment]
@IgorWojas this is a perfect place to use a [custom ensmallen callback](https://www.ensmallen.org/docs.html#custom-callbacks). I think this one should do about what you are hoping:

```c++
class PrintLossSometimes
{
 public:
  bool EndEpoch(OptimizerType& /* opt */,
                FunctionType& /* function */,
                const MatType& /* coordinates */,
                const size_t epoch,
                const double objective)
  {
    if (epochs % 1000 == 0) { cout << "loss = " << lossVal << "\n"; }
    return false; // Don't terminate optimization.
  }
};
```

Then just use `PrintLossSometimes()` instead of `PrintLoss()`. :+1:

**Comment by IgorWojas** [comment]
@vibhorjoshi I'm not sure how to do that...

**Comment by IgorWojas** [comment]
Before this gets closed... The solution for my issue was:

double LossMSE(const arma::mat& predicted, const arma::mat& target) { 
    return (accu(pow(target - predicted, 2))) / target.n_elem; }

double LossMAE(const arma::mat& predicted, const arma::mat& target) {
    return (accu(abs(target - predicted))) / target.n_elem;
}

Thanks All for your input
Cheers!!!

**Comment by rcurtin** [comment]
Awesome, glad you got it worked out!  I'll go ahead and close the issue then. :+1:
